connections/README.md

```diff
@@ -0,0 +1,341 @@
+back up" and rejoin a dense network between invocations of your libp2p
+application without having to rely on a few bootstrap nodes and random DHT walks
+to build up a routing table.
+
+Even during a single invocation of an application, you're likely to benefit from
+an in-memory metadata storage facility, which will allow you to cache addresses
+for connection resumption. Designing a storage interface which can be backed by
+memory or persistent storage will let you swap in whichever is appropriate for
+your use case and stage of development.
+
+For examples, see [go-libp2p-peerstore][go-libp2p-peerstore] and
+[js-peer-book][js-peer-book].
+
+#### Connection Limits
+
+Maintaining a large number of persistent connections can cause issues with some
+network environments and can lead to resource exhaustion and erratic behavior.
+
+It's highly recommended that libp2p implementations maintain an upper bound on
+the number of open connections. Doing so while still maintaining robust
+performance and connectivity will likely require implementing some kind of
+priority mechanism for selecting which connections are the most "expendable"
+when you're near the limit.
+
+Resource allocation, measurement and enforcement policies are all an active area
+of discussion in the libp2p community, and implementations are free to develop
+whatever prioritization system makes sense.
+
+There is currently a draft of a [connection manager
+specification][connmgr-v2-spec] that may replace the current [connmgr
+interface][connmgr-go-interface] in go-libp2p and may also form the basis of
+other connection manager implementations. 
+
+### Connection Lifecycle Events
+
+The establishment of new connections and streams is likely to be a
+"cross-cutting concern" that's of interest to various parts of your application
+(or parts of libp2p) besides the protocol handlers that directly deal with the
+traffic.
+
+For example, the [persistent metadata component](#peer-metadata-storage) could
+automatically add peer ids and addresses to its registry whenever a new peer
+connects, or a DHT module could update its routing tables when a connection is
+terminated.
+
+To support this, it's recommended that libp2p implementations support a
+notification or event delivery system that can inform interested parties about
+connection lifecycle events.
+
+The full set of lifecycle events is not currently specified, but a recommended
+baseline would be:
+
+| Event        | Description                               |
+|--------------|-------------------------------------------|
+| Connected    | A new connection has been opened          |
+| Disconnected | A connection has closed                   |
+| OpenedStream | A new stream has opened over a connection |
+| ClosedStream | A stream has closed                       |
+| Listen       | We've started listening on a new address  |
+| ListenClose  | We've stopped listening on an address     |
+
```

Comment:
in the future, we're eager to explore a services style model to allow subsystems to communicate. maybe doesn't need to be in this draft directly, but something to think about!